api/naming_service.go

```diff
@@ -27,40 +30,45 @@ var TLDMapping = map[string]uint64{}
 // @Router /ns/lookup [get]
 func MakeLookupRoute(router gin.IRouter) {
 	ns := router.Group("/ns")
-	ns.GET("/lookup", handleLookup)
+	ns.GET("/lookup", func(c *gin.Context) {
+		name := c.Query("name")
+		coinQuery := c.Query("coin")
+		coin, err := strconv.ParseUint(coinQuery, 10, 64)
+		if err != nil {
+			ginutils.RenderError(c, http.StatusBadRequest, "coin query is invalid")
+			return
+		}
+
+		result, err := handleLookup(name, coin)
+		if err != nil {
+			ginutils.RenderError(c, http.StatusBadRequest, err.Error())
+		}
+		ginutils.RenderSuccess(c, result)
+	})
 
 	TLDMapping[".eth"] = CoinType.ETH
 	TLDMapping[".xyz"] = CoinType.ETH
 	TLDMapping[".luxe"] = CoinType.ETH
 	TLDMapping[".zil"] = CoinType.ZIL
 }
 
-func handleLookup(c *gin.Context) {
-	name := c.Query("name")
-	coinQuery := c.Query("coin")
-
+func handleLookup(name string, coin uint64) (blockatlas.Resolved, error) {
+	result := blockatlas.Resolved{Result: "", Coin: coin}
 	if name == "" {
```

Comment:
can we move this verification to the top-level function? I think it's not the responsibility of this method to verify this. 

Just a suggestion

Reply:
Hm, do you think so? top level function just parses query data in this case. And this func is pure and checks if parameters passed to it are normal. It's also used for batch request.

Reply:
yes, The top level is for inputs, we get all we need to parse and pass to the method to do only the method need to do

Reply:
The function can be viewed as an individual component. Its author is best placed to define any preconditions and check them. Checking them outside presupposes the caller knows the preconditions which may not be the case. Also by placing them inside the function you're assured every call is checked. In function body usually perfomed checks on any post-conditions before leaving the function.

Reply:
Just the case: imagine you have function calcSquare(x, y). Who is responsible to check x and y are not negative?

Reply:
Bullshit man. Why you do something and on next step throw exception? If your function as individual component you shouldt check data in function. In design theory you should thing that your function live in perfect place.
Looked other languages: swift kotlin, java. You can say that function should get only non null values.

@Pantani have good suggestion. It's error check name in this function.

Reply:
@madcake Good design is to check parameters in every place before passing them? In our case we have to do it twice (for single coin handler and for batch). 
Do you think we have to delegate checking functions parameters logic for somebody's else? Function that handles api call have no idea what is going to be done in the `handleLookup` but anyway has to somehow validate its parameters instead of just to be transparent and pass them through? What if it will be possible to pass only names prefixed with 'joe'. Do you think we have to handle this condition in every place somewhere but not in our func?

Why standard GO libraries return errors then? For example json.Unmarshal. Based on what you've just said we have to check that we pass valid json there by ourselves every time.

Reply:
@dpereskokov The function doesn't need to verify the params, this is not a good pattern in every language, we need to verify the inputs at a high level. When we receive these inputs, the functions need to do only the thing to in charge, not to verify the parameters. Verification needs to be done in the parse

Reply:
> Just the case: imagine you have function calcSquare(x, y). Who is responsible to check x and y are not negative?

If the software is only to do that, make sense, but it is only a method in the code, and do you need to verify always do call this method, you gonna do a lot of duplicated code

The code needs to return an error and we handle the error. This is the reason the Go is oriented to errors. `calcSquare(x, y) (float64, error)`

Reply:
param validation inside the method needs to be an exception. The input always need to be verified at the top level, and the low levels need to do only the things they born to do